### ISSUE TITLE
Format JSON for diff compatibility?

### ISSUE THREAD

**Issue by TravisCardwell** [mzabani/codd#137]
Codd represents the state of a database at a given time using JSON, either organized hierarchically by key using many files or in a single JSON file.  In both cases, the JSON is optimized for small size by minimizing whitespace.  Example:

```json
{"collate":"en_us.utf8","ctype":"en_us.utf8","encoding":"utf8","owner":"postgres","public_privileges":"[[\"cT\", \"postgres\"]]","settings":[{"enum_vals":null,"max_val":null,"min_val":null,"name":"default_transaction_deferrable","value":"off"},{"enum_vals":["read committed","read uncommitted","repeatable read","serializable"],"max_val":null,"min_val":null,"name":"default_transaction_isolation","value":"read committed"},{"enum_vals":null,"max_val":null,"min_val":null,"name":"default_transaction_read_only","value":"off"}]}
```

When I inspect the state, I have *always* pretty-printed the JSON so that I can read it easily as well as use [`diff`][] to inspect differences.  Specifically, pretty-printed JSON puts each property on a different line, so it works well with line-based comparison software.  Pretty-printed version of the above example:

[`diff`]: <https://linux.die.net/man/1/diff>

```json
{
    "collate": "en_us.utf8",
    "ctype": "en_us.utf8",
    "encoding": "utf8",
    "owner": "postgres",
    "public_privileges": "[[\"cT\", \"postgres\"]]",
    "settings": [
        {
            "enum_vals": null,
            "max_val": null,
            "min_val": null,
            "name": "default_transaction_deferrable",
            "value": "off"
        },
        {
            "enum_vals": [
                "read committed",
                "read uncommitted",
                "repeatable read",
                "serializable"
            ],
            "max_val": null,
            "min_val": null,
            "name": "default_transaction_isolation",
            "value": "read committed"
        },
        {
            "enum_vals": null,
            "max_val": null,
            "min_val": null,
            "name": "default_transaction_read_only",
            "value": "off"
        }
    ]
}
```

Might it be preferable to pretty-print JSON (using [`aeson-pretty`][])?

[`aeson-pretty`]: <https://hackage.haskell.org/package/aeson-pretty>

Pros:

* Users can easily use [`diff`][] software to compare different states.  When using Git, this includes `diff` views using [`git diff`][], GitHub, GitLab, other Git web applications, desktop GUI software, editors, etc.
* Users can easily inspect the state without having to pipe it through a tool to pretty-print it.

[`git diff`]: <https://git-scm.com/docs/git-diff>

Cons:

* Additional whitespace makes pretty-printed JSON larger than minimized JSON.  The following examples use two-space indentation in the pretty-printed JSON.

    ```
    $ cat db-settings | wc -c
    526
    $ cat db-settings | python -m json.tool --indent 2 | wc -c
    748
    $ cat db-settings | gzip | wc -c
    237
    $ cat db-settings | python -m json.tool --indent 2 | gzip | wc -c
    268
    ```

    ```
    $ cat codd-state.json | wc -c
    181915
    $ cat codd-state.json | python -m json.tool --indent 2 | wc -c
    373174
    $ cat codd-state.json | gzip | wc -c
    12412
    $ cat codd-state.json | python -m json.tool --indent 2 | gzip | wc -c
    16248
    ```

**Comment by mzabani** [comment]
This thought has occurred to me, but the reason why json representations are single line is because that makes it easier to detect conflicting changes across different PRs as merge conflicts.

While db-settings (the example you used here) looks like it'd benefit from pretty printing without any obvious downsides (and I'm sure many other objects might fall into that same category), take for example a column:

```json
{
  "collation": "default",
  "collation_nsp": "pg_catalog",
  "default": null,
  "generated": "",
  "hasdefault": false,
  "identity": "",
  "inhcount": 0,
  "local": true,
  "notnull": true,
  "order": 2,
  "privileges": null,
  "type": "text"
}
```

Say one developer changes the default value of this column to `'default-string'` and another developer changes its type to `numeric`: if representations are single lines this will be a merge conflict, but if they're not it won't.

Admittedly, merge conflicts will still exist if we pretty-print representations; they'll just be more precise. The current method errs in the side of more conflicts, including more false positives.

Also, even a trivial test such as running `codd up` in CI should catch the example above, so merge conflicts serving as earlier signals of possible problems is not that great a feature in many cases. Plus, I don't know how to gather data to judge what's best in the spectrum of more/less false positives.

However, I do tend to think (without data to back it up) that if two developers modify the same object, it is best one of them **knows** about it through a merge conflict. One more silly example to back this up is a developer changing a column's order and the other writing a new `SELECT *` query in a new PR: even tests might not detect this if the types of columns in the new ordering match the previous one.

Again I recognize the example above is very contrived, and to defend pretty printing a little, I don't think anyone would blame codd for not forcing a merge conflict in cases such as the ones I exemplified, though I wouldn't be surprised if they thought about the opposite idea: "maybe codd should render one-line representations so these conflicts are shown to us".

What do you think?

**Comment by TravisCardwell** [comment]
That is a very good point.  I also prefer to catch conflicts earlier.  It is a difficult trade-off, however.  Since some conflicts can only be caught by tests regardless and rendering JSON as one line likely results in more false positives than actual conflicts, my personal preference is to prioritize readable `diff`s.

I experimented a bit with migration conflicts and found that it is more complicated than I had anticipated ([blog entry](https://www.extrema.is/blog/2023/04/06/codd-experiment-4-migration-conflicts)).  The example that I used in the blog is that of two migrations adding columns to the same table.  Git does not register a conflict, but automatically merged schema information is invalid due to conflicting column order.  In this case, the conflicting values are in different files.

A migration that conflicts with a query is a great example of a type of conflict that can only be caught using tests!  (One should avoid using `SELECT *` in general, but other types of queries could indeed be incompatible with a migration.)

Perhaps it would be beneficial to be notified when merging any new migrations.  I use indexes to organize migrations.  This is not necessary with the Codd style of managing migrations because Codd migrations are added to a managed directory and timestamped.  I manage inverse migrations, multi-step migrations, and checkpoint migrations, however, and not all of the migrations are added.  (Inverse migrations should never be added!)  The indexes are required to keep all of these migrations organized.  When merging new migrations, the desired order needs to be determined and the indexes updated accordingly even if there are no conflicts in the resulting state.  I have not implemented it yet, but my idea for this is to persist a reference to the last migration so that Git registers a conflict in this case.

**Comment by mzabani** [comment]
> I experimented a bit with migration conflicts and found that it is more complicated than I had anticipated ([blog entry](https://www.extrema.is/blog/2023/04/06/codd-experiment-4-migration-conflicts)). The example that I used in the blog is that of two migrations adding columns to the same table. Git does not register a conflict, but automatically merged schema information is invalid due to conflicting column order. In this case, the conflicting values are in different files.

Thanks for the articles, and that's an interesting example. It reminds me that it has been asked before that all columns go into a single file in the on-disk representation of codd. The reasoning was different then: removing a column can make the git diff touch _lots_ of files, making reviews more painful in platforms like github.

But I think pretty-printing an array of column representations into a single file would also make your example a merge conflict?

In any case, I think I've been convincing myself that slightly-earlier meaningful merge conflicts are hard to come up with and aren't worth the loss of readability; I can't remember one at work to begin with, and the examples I gave earlier are all very contrived. At least if we pretty-print json we'll be able to say conflicts are decidedly in the "few false-positives" end of the spectrum instead of having to decide on a case by case basis.

> The indexes are required to keep all of these migrations organized. When merging new migrations, the desired order needs to be determined and the indexes updated accordingly even if there are no conflicts in the resulting state. I have not implemented it yet, but my idea for this is to persist a reference to the last migration so that Git registers a conflict in this case.

I think I'm starting to better understand your use case. You deploy to environments in a blue-green-safe style, i.e. with non-destructive and destructive migrations applied in separate deployments, and you need to worry about undoing migrations in case some specific state of the database didn't fare well with the application, after all.

To be able to undo things, it would help you to be able to keep a single representation/hash of DB state that you track for each migration and can refer to, so that you could e.g. `codd revert-to some-working-db-state-hash` if things go bad. That comes with that extra management of indexing migrations numerically, which means re-generating whole-DB state if another developer's migration is merged before yours.

Am I getting closer to understanding your use case?

**Comment by TravisCardwell** [comment]
> Thanks for the articles, and that's an interesting example. It reminds me that it has been asked before that all columns go into a single file in the on-disk representation of codd. The reasoning was different then: removing a column can make the git diff touch _lots_ of files, making reviews more painful in platforms like github.

One way to resolve this issue is to put the column order in the table configuration (`objrep`) instead of each column configuration.  Example:

```json
{
  ...
  "column_order": ["id", "name", "email"],
  ...
}
```

> But I think pretty-printing an array of column representations into a single file would also make your example a merge conflict?

Indeed!  With a representation like the above, any simultaneous addition or removal of columns in a table should result in a merge conflict.

> In any case, I think I've been convincing myself that slightly-earlier meaningful merge conflicts are hard to come up with and aren't worth the loss of readability; I can't remember one at work to begin with, and the examples I gave earlier are all very contrived. At least if we pretty-print json we'll be able to say conflicts are decidedly in the "few false-positives" end of the spectrum instead of having to decide on a case by case basis.

I guess it is also possible to allow users to configure the behavior...  I am not sure if it is worth it, though.

```bash
export CODD_JSON=pretty
```

> I think I'm starting to better understand your use case. You deploy to environments in a blue-green-safe style, i.e. with non-destructive and destructive migrations applied in separate deployments, and you need to worry about undoing migrations in case some specific state of the database didn't fare well with the application, after all.
>
> To be able to undo things, it would help you to be able to keep a single representation/hash of DB state that you track for each migration and can refer to, so that you could e.g. `codd revert-to some-working-db-state-hash` if things go bad.
> ...
>
> Am I getting closer to understanding your use case?

Yes, I think so.

Note that reverting state using inverse migrations is sometimes required for reasons that are unrelated to the database.  For example, a migration may be part of an upgrade that needs to be rolled back due to an issue with a separate service that only happens in production.  (Developers should separate potentially problematic upgrades when possible, of course!)

In the design ideas, a list of applied migrations is not maintained because different databases may be at different states of the migration digraph, and they may have used different migrations to reach the same state.  The migration software maintains state that represents the migration digraph, using hashes to identify states.  Each database tracks all "transitions" (migration application, database upgrade, migration software upgrade) that have been performed.  If the latest recorded database (major) version or migration software (major) version differs from the current versions, then appropriate actions can be taken.  The recorded current state hash can be used to verify the current state without any "snapshots," while an existing snapshot (or on-demand snapshot) can be used to provide detailed `diff` information.

With this design, migrations to apply can be determined by specifying a migration.  For example, consider the following digraph, where the inverses are not labeled just because Mermaid does not support TeX syntax:

```mermaid
graph LR;
  A-- m1 -->B; B-->A;
  B-- m2 -->C; C-->B;
  C-- m3 -->D; D-->C;
  D-- m4 -->E; E-->D;
```

If a database is in state `C`, then command `codd database migrate 0004-m4.sql` would do the following:

1. Determine the current state of the database (`C`).
2. Determine the (shortest) path in the digraph from the current state (`C`) that ends with the specified migration (`m4`).
3. Apply migration `m3`, transitioning the database from state `C` to state `D`.
4. Apply migration `m4`, transitioning the database from state `D` to state `E`.

State verification could be done before/after after step, only at the beginning and end, or only at the end, according to user options.

Inverse migrations are special only in terms of defining the digraph and verifying the properties that inverses must have.  Applying inverse migrations would use the same command.  Command `codd database migrate 0002-m2.inverse.sql` would do the following:

1. Determine the current state of the database (`C`).
2. Determine the (shortest) path in the digraph from the current state (`C`) that ends with the specified migration (`m2` inverse).
3. Apply migration `m2`inverse, transitioning the database from state `C` to state `B`.

The migration argument could default to the "last" migration, in order to have behavior similar to `codd up`.

> That comes with that extra management of indexing migrations numerically, which means re-generating whole-DB state if another developer's migration is merged before yours.

Indexing is required in order to organize the different types of migration files as well as define the migration digraph.

Regenerating the DB state is a general issue:

* With Codd, migration conflicts often involve regenerating the DB state as well as rebuilding the "snapshot" (`expected-schema`).  I use `git checkout HASH ...` to checkout snapshots from which I can add corrected migrations.
* With the above design, one can either transition to a previous state using an inverse or regenerate the DB state.  Since all snapshots are stored using hashes, there are never Git-style conflicts in snapshots.

**Comment by mzabani** [comment]
I think I get it. Thanks for explaining and even including diagrams.

I'm not sure I understand if regenerating DB state is a general issue, though. With codd as it currently is, you only need to regenerate DB state in some cases, but many go unnoticed. With migrations as edges of a digraph, any new migration will need to have its vertices (which are DB states) regenerated every time a different migration was merged before yours, even if they modify completely unrelated DB objects.

However, even if adding a migration becomes a strictly serialized operation for developers, I don't think it's a big deal (it can even be automated in many cases) and I see a lot of value in this mode of operation for gradual rollouts of functionality, as I'm sure you know better than me. Thanks a lot for coming up with this; it is a simple model (one I can fit in my small head easily) and yet it seems very powerful for these more complex use cases.

I think we could very well study how to fit this into codd. IIUC you "simply" have to assign numbers to migrations to identify order (as in [your article](https://www.extrema.is/blog/2023/04/03/codd-migration-inverses)) and store from-state and to-state in each migration (even if we could infer order from each migration's from-state and to-state, we might as well number them to make it easy on us poor humans IIUC). In fact, I don't need to say this as you already know it and your article does a very good job at explaining it.

I fear we've diverted from the original issue, though. Maybe we should create a new one and dedicate it to fitting your model into codd? I think it might be achievable if codd branches its behaviour early during execution by realising migrations are not timestamped (though maybe an explicit flag during invocation or env var might be an option).

We could use codd's special comments to encode from-state and to-state, and even have a `-- codd: inverse` section that you could put in the same sql file as the forward migration. I used to have that for what I called blue-green-safe migrations and it was tricky to maintain, but that's just because control flow and the streaming parser were a tough couple to marry at the codebase at the time; it should be much easier now, hopefully.

One concern I have is how to encode full database state in a backwards compatible manner. Say that from-state and to-state are hashes of codd's json representations (but this works even if they're the result of some function of json representations). Then even trivial bugfixes like the one for https://github.com/mzabani/codd/issues/140 can change the values of previous states, right? Have you thought about this?

-------------------------------------

For this issue, though, I think I'll try a single file `columns` with an array of column representations.

> I guess it is also possible to allow users to configure the behavior... I am not sure if it is worth it, though.

I agree it's not worth it. Let's make all json pretty.

> One way to resolve this issue is to put the column order in the table configuration (objrep) instead of each column configuration.

That might be worth a try, although if I go with a single `columns` file it seems this would be redundant, and the single `columns` file has the advantage of being easier to review in platforms like github. But if I'm missing something about this particular representation, please let me know!

**Comment by TravisCardwell** [comment]
@mzabani Sorry for the long delay in my replies!  I have not had a chance to work on Codd for quite some time.  This week, I kept thinking that I would get a chance to work on it "after a little bit" but did not.  I hope to reply next week! :bow:

**Comment by mzabani** [comment]
Don't worry, that's just life! Please take your time, there is no hurry.

**Comment by TravisCardwell** [comment]
Moving the design discussion to a separate issue sounds good to me.  For this comment, I will use a horizontal rule.

Putting inverses in the same files as the migrations, delimiting them using special syntax (`-- codd: inverse`), would decrease the number of migration files.  In the draft design, a filename is sufficient to identify an arrow, however, while such a design would require additionally indicating the part of the file.  Specifying an arrow with (just) a filename is convenient because command-line completion can be used.  Another benefit of separate files is that they can be used even without Codd, making them more versatile for testing or manually dealing with issues if/when something goes wrong.

Indeed, changes in Codd can change the way that database state is represented.  This is the reason why my draft design tracks both the software (major) version and the database (major) version.  The hash that represents a state is specific to both.  In my prototype, the JSON file that stores the state information of the migration digraph tracks state hashes per software (major) version and database (major) version, allowing correlation.  It effectively adds dimensions to the digraph, but that sounds more complicated than it actually is.

Wow, it has been more than a month since I have had time to work on this.  There are some tasks that must be done before I can continue with migration efforts in the project that I am working on.  Thank you for your understanding!  C'est la vie.

----

Putting column representations in a single `columns` file is an interesting idea.  Using an array would specify the order, without having to to use an explicit property.  My primary concern is with Git automatic merges.  Does it work correctly when similar columns are added?  I suspect that it could do the wrong thing without flagging a conflict, in some cases, but I have not tested it.

**Comment by mzabani** [comment]
> Specifying an arrow with (just) a filename is convenient because command-line completion can be used. Another benefit of separate files is that they can be used even without Codd, making them more versatile for testing or manually dealing with issues if/when something goes wrong.

Ah, you have convinced me completely. Code will also be simpler, so another +1.

----------------------------------------------------

> My primary concern is with Git automatic merges. Does it work correctly when similar columns are added? I suspect that it could do the wrong thing without flagging a conflict, in some cases, but I have not tested it.

I'm not familiar with the innards of git's merging algorithm, but we can do bigger changes like merging columns into a single file as a separate issue, and just pretty-print representations as they currently are first. Then this can be tested to see how git would handle it.
